File: backend/parsers/starling_csv.py
"""Starling Bank CSV statement parser."""
import csv
from datetime import datetime, date
from io import StringIO


class StarlingCSVParser:
    """Parser for Starling Bank CSV exports.

    CSV Format:
    Date,Counter Party,Reference,Type,Amount (GBP),Balance (GBP),Spending Category,Notes
    01/11/2025,Tesco,TESCO METRO,APPLE PAY,-7.10,18.06,GROCERIES,
    """

    # Map Starling categories to our category names
    CATEGORY_MAP = {
        "GROCERIES": "Groceries",
        "EATING_OUT": "Food & Dining",
        "ENTERTAINMENT": "Entertainment",
        "TRANSPORT": "Transport",
        "SHOPPING": "Shopping",
        "BILLS": "Utilities",
        "INCOME": "Income",
        "TRANSFERS": "Transfer In",
        "GENERAL": "Other",
        "LIFESTYLE": "Personal Care",
        "HOLIDAYS": "Travel",
        "FAMILY": "Other",
        "CHARITY": "Other",
        "GAMBLING": "Entertainment",
        "SAVINGS": "Transfer Out",
        "PAYMENTS": "Transfer Out",
    }

    def parse(self, content: bytes) -> dict:
        """
        Parse Starling CSV statement.

        Args:
            content: CSV file bytes

        Returns:
            Dict with keys:
                - period_start: date
                - period_end: date
                - opening_balance: float | None
                - closing_balance: float | None
                - transactions: list[dict]
                - raw_text: str
        """
        # Decode CSV
        text = content.decode("utf-8-sig")  # Handle BOM if present
        reader = csv.DictReader(StringIO(text))

        transactions = []
        dates = []
        balances = []

        for row in reader:
            try:
                # Parse date (DD/MM/YYYY)
                txn_date = datetime.strptime(row["Date"], "%d/%m/%Y").date()
                dates.append(txn_date)

                # Parse amount
                amount = float(row["Amount (GBP)"].replace(",", ""))

                # Parse balance
                balance = float(row["Balance (GBP)"].replace(",", ""))
                balances.append(balance)

                # Build description
                counter_party = row.get("Counter Party", "").strip()
                reference = row.get("Reference", "").strip()
                txn_type = row.get("Type", "").strip()

                if reference and reference != counter_party:
                    description = f"{counter_party} - {reference}"
                else:
                    description = counter_party

                if txn_type:
                    description = f"{description} ({txn_type})"

                # Get Starling category
                starling_category = row.get("Spending Category", "").strip()
                mapped_category = self.CATEGORY_MAP.get(starling_category)

                transactions.append({
                    "date": txn_date,
                    "description": description,
                    "amount": amount,
                    "balance": balance,
                    "starling_category": starling_category,
                    "mapped_category": mapped_category,
                    "notes": row.get("Notes", "").strip() or None,
                })

            except (KeyError, ValueError) as e:
                # Skip malformed rows
                continue

        # Determine period and balances
        if dates:
            period_start = min(dates)
            period_end = max(dates)
        else:
            period_start = period_end = date.today()

        # Opening balance: first transaction's balance - first transaction's amount
        # Closing balance: last transaction's balance
        opening_balance = None
        closing_balance = None

        if transactions:
            # Sort by date while keeping original order for same-day items
            transactions = [
                txn for _, txn in sorted(
                    enumerate(transactions),
                    key=lambda pair: (pair[1]["date"], pair[0]),
                )
            ]

            # Opening = balance before first transaction
            first_txn = transactions[0]
            opening_balance = first_txn["balance"] - first_txn["amount"]

            # Closing = last balance
            closing_balance = transactions[-1]["balance"]

        return {
            "period_start": period_start,
            "period_end": period_end,
            "opening_balance": opening_balance,
            "closing_balance": closing_balance,
            "transactions": transactions,
            "raw_text": text,
        }

File: backend-go/internal/parsers/starling_csv.go
package parsers

import (
	"encoding/csv"
	"fmt"
	"io"
	"sort"
	"strconv"
	"strings"
	"time"
)

// StarlingCSVParser parses Starling Bank CSV exports.
//
// CSV Format:
// Date,Counter Party,Reference,Type,Amount (GBP),Balance (GBP),Spending Category,Notes
type StarlingCSVParser struct{}

// Starling category -> our category name
var starlingCategoryMap = map[string]string{
	"GROCERIES":     "Groceries",
	"EATING_OUT":    "Food & Dining",
	"ENTERTAINMENT": "Entertainment",
	"TRANSPORT":     "Transport",
	"SHOPPING":      "Shopping",
	"BILLS":         "Utilities",
	"INCOME":        "Income",
	"TRANSFERS":     "Transfer In",
	"GENERAL":       "Other",
	"LIFESTYLE":     "Personal Care",
	"HOLIDAYS":      "Travel",
	"FAMILY":        "Other",
	"CHARITY":       "Other",
	"GAMBLING":      "Entertainment",
	"SAVINGS":       "Transfer Out",
	"PAYMENTS":      "Transfer Out",
}

func (p *StarlingCSVParser) FileTypes() []string { return []string{".csv"} }
func (p *StarlingCSVParser) BankID() string      { return "starling" }
func (p *StarlingCSVParser) BankName() string     { return "Starling Bank" }

func (p *StarlingCSVParser) Parse(content []byte) (ParsedStatement, error) {
	// Strip BOM
	text := string(content)
	text = strings.TrimPrefix(text, "\xef\xbb\xbf")

	r := csv.NewReader(strings.NewReader(text))
	header, err := r.Read()
	if err != nil {
		return ParsedStatement{}, fmt.Errorf("read header: %w", err)
	}

	colIdx := map[string]int{}
	for i, h := range header {
		colIdx[strings.TrimSpace(h)] = i
	}

	type indexedTxn struct {
		idx int
		txn ParsedTransaction
	}

	var txns []indexedTxn
	var dates []time.Time
	idx := 0

	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			continue
		}

		dateStr := getCol(row, colIdx, "Date")
		txnDate, err := time.Parse("02/01/2006", dateStr)
		if err != nil {
			continue
		}
		dates = append(dates, txnDate)

		amountStr := strings.ReplaceAll(getCol(row, colIdx, "Amount (GBP)"), ",", "")
		amount, err := strconv.ParseFloat(amountStr, 64)
		if err != nil {
			continue
		}

		balanceStr := strings.ReplaceAll(getCol(row, colIdx, "Balance (GBP)"), ",", "")
		balance, err := strconv.ParseFloat(balanceStr, 64)
		if err != nil {
			continue
		}

		counterParty := strings.TrimSpace(getCol(row, colIdx, "Counter Party"))
		reference := strings.TrimSpace(getCol(row, colIdx, "Reference"))
		txnType := strings.TrimSpace(getCol(row, colIdx, "Type"))

		var description string
		if reference != "" && reference != counterParty {
			description = counterParty + " - " + reference
		} else {
			description = counterParty
		}
		if txnType != "" {
			description = description + " (" + txnType + ")"
		}

		starlingCat := strings.TrimSpace(getCol(row, colIdx, "Spending Category"))
		var mappedCat *string
		if mc, ok := starlingCategoryMap[starlingCat]; ok {
			mappedCat = &mc
		}

		notes := strings.TrimSpace(getCol(row, colIdx, "Notes"))
		var notesPtr *string
		if notes != "" {
			notesPtr = &notes
		}

		bal := balance
		txns = append(txns, indexedTxn{
			idx: idx,
			txn: ParsedTransaction{
				Date:           txnDate,
				Description:    description,
				Amount:         amount,
				Balance:        &bal,
				MappedCategory: mappedCat,
				Notes:          notesPtr,
			},
		})
		idx++
	}

	// Sort by date, preserving order within same day
	sort.SliceStable(txns, func(i, j int) bool {
		if txns[i].txn.Date.Equal(txns[j].txn.Date) {
			return txns[i].idx < txns[j].idx
		}
		return txns[i].txn.Date.Before(txns[j].txn.Date)
	})

	result := ParsedStatement{RawText: text}

	if len(dates) > 0 {
		minDate, maxDate := dates[0], dates[0]
		for _, d := range dates[1:] {
			if d.Before(minDate) {
				minDate = d
			}
			if d.After(maxDate) {
				maxDate = d
			}
		}
		result.PeriodStart = minDate
		result.PeriodEnd = maxDate
	} else {
		now := time.Now()
		result.PeriodStart = now
		result.PeriodEnd = now
	}

	sorted := make([]ParsedTransaction, len(txns))
	for i, t := range txns {
		sorted[i] = t.txn
	}

	if len(sorted) > 0 {
		opening := *sorted[0].Balance - sorted[0].Amount
		result.OpeningBalance = &opening
		closing := *sorted[len(sorted)-1].Balance
		result.ClosingBalance = &closing
	}

	result.Transactions = sorted
	return result, nil
}

func getCol(row []string, colIdx map[string]int, name string) string {
	if i, ok := colIdx[name]; ok && i < len(row) {
		return row[i]
	}
	return ""
}
